core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalLogWriter.java

```diff
@@ -149,24 +164,132 @@ public void close() throws IOException {
-          .writeDelimitedTo(mJournalOutputStream.mOutputStream);
+      JournalEntry entryToWrite =
+          entry.toBuilder().setSequenceNumber(mNextSequenceNumber).build();
+      entryToWrite.writeDelimitedTo(mJournalOutputStream.mOutputStream);
+      LOG.debug("Add this journal entry to retryList with {} entries.", mEntriesToFlush.size());
+      mEntriesToFlush.add(entryToWrite);
+      mNextSequenceNumber++;
     } catch (IOException e) {
+      // Set mJournalOutputStream to null so that {@code maybeRecoverFromUfsFailures}
+      // can know a UFS failure has occurred.
       mRotateLogForNextWrite = true;
+      UfsJournalFile currentLog = mJournalOutputStream.mCurrentLog;
+      mJournalOutputStream = null;
       throw new IOException(ExceptionMessage.JOURNAL_WRITE_FAILURE
           .getMessageWithUrl(RuntimeConstants.ALLUXIO_DEBUG_DOCS_URL,
-              mJournalOutputStream.mCurrentLog, e.getMessage()), e);
+              currentLog, e.getMessage()), e);
+    }
+  }
+
+  /**
+   * Core logic of UFS journal recovery from UFS failures.
+   *
+   * If Alluxio stores its journals in UFS, then Alluxio needs to handle UFS failures.
+   * When UFS is dead, there is nothing Alluxio can do because Alluxio relies on UFS to
+   * persist journal entries. Consequently any metadata operation will block because Alluxio
+   * cannot flush their journal entries.
+   * Once UFS comes back online, Alluxio needs to perform the following operations:
+   * 1. Locate the most recent incomplete journal file, i.e. journal file that starts with
+   *    a valid sequence number X (hex), and ends with 0x7fffffffffffffff. The journal file
+   *    name encodes this information, i.e. X-0x7fffffffffffffff.
+   * 2. Sequentially scan all the incomplete journal file, and identify the last valid journal
+   *    entry that has been persisted in UFS. Once we know the last valid journal entry in UFS,
+   *    we know the first journal entry that is not valid in UFS. Suppose it is Y.
+   * 3. Rename the incomplete journal file to X-Y. Future journal writes will write to a new file
+   *    named Y-0x7fffffffffffffff.
+   * 4. Check whether there is any missing journal entry between Y and the oldest entry in
+   *    mEntriesToFlush, say Z. If Z > Y, then it means journal entries in [Y, Z) are
+   *    missing, and Alluxio cannot recover. Otherwise, for each journal entry in
+   *    mEntriesToFlush, if its sequence number is newer than or equal to Y, retry writing it to
+   *    UFS by calling the {@code UfsJournalLogWriter#write} method.
+   * @throws IOException if reading journal entries from UFS fails
+   */
+  private void maybeRecoverFromUfsFailures() throws IOException {
+    if (mJournalOutputStream != null) {
+      return;
+    }
+    UfsJournalSnapshot snapshot = UfsJournalSnapshot.getSnapshot(mJournal);
+    UfsJournalFile currentLog = snapshot.getCurrentLog(mJournal);
+    if (currentLog == null) {
+      return;
```

Comment:
when would this happen?

Reply:
This can happen when
1. the UFS journal directory does not have any journal file. Or
2. the UFS journal directory does not have incomplete journal file.
1 may happen if UFS fails before Alluxio has a chance to persist any journal entry.
2 may happen if the incomplete journal file is missing.

Reply:
got it, thanks

Reply:
Done.

Reply:
Done.